Source/SwiftLintFramework/Rules/Lint/DuplicateConditionsRule.swift

```diff
@@ -0,0 +1,227 @@
+import SwiftSyntax
+
+struct DuplicateConditionsRule: SwiftSyntaxRule, ConfigurationProviderRule {
+    var configuration = SeverityConfiguration(.warning)
+
+    init() {}
+
+    static let description = RuleDescription(
+        identifier: "duplicate_conditions",
+        name: "Duplicate Conditions",
+        description: "Duplicate sets of conditions in the same branch instruction should be avoided",
+        kind: .lint,
+        nonTriggeringExamples: [
+            Example("""
+                if x < 5 {
+                  foo()
+                } else if y == "s" {
+                  bar()
+                }
+            """),
+            Example("""
+                if x < 5 {
+                  foo()
+                }
+                if x < 5 {
+                  bar()
+                }
+            """),
+            Example("""
+                if x < 5, y == "s" {
+                  foo()
+                } else if x < 5 {
+                  bar()
+                }
+            """),
+            Example("""
+                switch x {
+                case \"a\":
+                  foo()
+                  bar()
+                }
+            """),
+            Example("""
+                switch x {
+                case \"a\" where y == "s":
+                  foo()
+                case \"a\" where y == "t":
+                  bar()
+                }
+            """),
+            Example("""
+                if let x = maybeAbc {
+                  foo()
+                } else if let x = maybePqr {
+                  bar()
+                }
+            """),
+            Example("""
+                if let x = maybeAbc, let z = x.maybeY {
+                  foo()
+                } else if let x = maybePqr, let z = x.maybeY {
+                  bar()
+                }
+            """),
+            Example("""
+                if case .p = x {
+                  foo()
+                } else if case .q = x {
+                  bar()
+                }
+            """)
+        ],
+        triggeringExamples: [
+            Example("""
+                if ↓x < 5 {
+                  foo()
+                } else if y == "s" {
+                  bar()
+                } else if ↓x < 5 {
+                  baz()
+                }
+            """),
+            Example("""
+                if z {
+                  if ↓x < 5 {
+                    foo()
+                  } else if y == "s" {
+                    bar()
+                  } else if ↓x < 5 {
+                    baz()
+                  }
+                }
+            """),
+            Example("""
+                if ↓x < 5, y == "s" {
+                  foo()
+                } else if x < 10 {
+                  bar()
+                } else if ↓y == "s", x < 5 {
+                  baz()
+                }
```

Comment:
Is it also possible to mix `,` and `&&` and still get the duplicated condition detected?

```swift
if x < 5, y == "s" {
  foo()
} else if x < 10 {
  bar()
} else if y == "s" && x < 5 {
  baz()
}
```

Reply:
So, in theory, yes. But I think doing it right makes the implementation much, much trickier

When we have `x < 5, y == "s"`, the syntax tree pulls these apart for us as distinct `SequenceExprSyntax` nodes. We can easily process them separately. If we have `x < 5 && y == "s"`, this is now a single `SequenceExprSyntax` node, with an `ExprListSyntax` with a bunch of children of differing types. One of these is the `BinaryOperatorExprSyntax` we want, the `&&` operator, so we'd have to break apart the list on those operators, and look at those individually:
```
0: SequenceExprSyntax children=1
  0: ExprListSyntax children=7
    0: IdentifierExprSyntax children=1
      0: identifier("y")
    1: BinaryOperatorExprSyntax children=1
      0: binaryOperator("==")
    2: StringLiteralExprSyntax children=3
      0: stringQuote
      1: StringLiteralSegmentsSyntax children=1
        0: StringSegmentSyntax children=1
          0: stringSegment("s")
      2: stringQuote
    3: BinaryOperatorExprSyntax children=1
      0: binaryOperator("&&")
    4: IdentifierExprSyntax children=1
      0: identifier("x")
    5: BinaryOperatorExprSyntax children=1
      0: binaryOperator("<")
    6: IntegerLiteralExprSyntax children=1
      0: integerLiteral("5")
```


But then what happens if we wrap in brackets? If we instead have `if (x < 5 && y == "s")`, then even though this is "the same", we now have a `TupleExprSyntax` wrapping the original `SequenceExprSyntax`. So we'd have to figure out that we need to drill down through these as many times as we need, before figuring out the `&&`'s. I would imagine there are also concerns about operator precedence that would need to be accounted for (in particular, note the `==` is at the same level as the `&&` in the above breakdown; is it obvious how to split up `p && q || r && s`?), and that there could be issues I haven't thought of...

I don't know this codebase, so if you know of any tools that already exist to help with this complexity, I'm very open to trying this out! But otherwise, I think limiting the scope to ensure it's correct might be safer. Thoughts?

Reply:
You are totally right. I don’t know if there is any easy way to achieve this already, since I’m also new to the codebase.
I think this feature would be nice to have, but for an first iteration it is probably not needed. Maybe we can create a follow-up issue to track this feature.

Reply:
> I don't know this codebase, so if you know of any tools that already exist to help with this complexity, I'm very open to trying this out! But otherwise, I think limiting the scope for now to ensure it's correct might be safer. Thoughts?

I agree that the current scope is sufficient as a first implementation. However, the functionality you are looking for, that is converting an expression list into a syntax tree taking precedence rules into account, already exists. See `IdenticalOperandsRule` as an example on how this works.